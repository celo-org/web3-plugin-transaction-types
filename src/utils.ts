import { registryABI, feeCurrencyWhitelistABI } from "@celo/abis";
import { feeCurrencyDirectoryABI } from "@celo/abis-l2";
import Web3, { Address, Contract } from "web3";
import { CeloTransactionTypesPlugin } from ".";

export enum TxTypeToPrefix {
  "ethereum-legacy" = "",
  cip42 = "0x7c",
  cip64 = "0x7b",
  cip66 = "0x7a",
  eip1559 = "0x02",
}

const celo = {
  chainId: 42220,
  rpcUrl: "https://forno.celo.org",
  explorerUrl: "https://celoscan.io",
} as const;
const alfajores = {
  chainId: 44787,
  rpcUrl: "https://alfajores-forno.celo-testnet.org",
  explorerUrl: "https://alfajores.celoscan.io",
} as const;
const dango = {
  chainId: 44787,
  rpcUrl: "https://forno.dango.celo-testnet.org",
  explorerUrl: "https://celo-dango.blockscout.com/",
} as const;

export const CeloChains = {
  celo,
  mainnet: celo,
  alfajores,
  testnet: alfajores,
  dango,
  testnetl2: dango,
} as const;

export const PROXY_ADMIN_ADDRESS = "0x4200000000000000000000000000000000000018";
export const REGISTRY_ADDRESS = "0x000000000000000000000000000000000000ce10";

export const getContractAddress = async (
  ctx: CeloTransactionTypesPlugin,
  contractName: string
) => {
  const contract = new Contract(registryABI, REGISTRY_ADDRESS);
  contract.link(ctx);

  const address = await contract.methods
    .getAddressForString(contractName)
    .call();

  return address;
};

export const isCel2 = async (ctx: CeloTransactionTypesPlugin) => {
  const web3 = new Web3(ctx.provider);
  const code = await web3.eth.getCode(PROXY_ADMIN_ADDRESS);

  if (typeof code === "string") {
    return code != "0x" && code.length > 2;
  }

  return false;
};

export async function isWhitelisted(
  ctx: CeloTransactionTypesPlugin,
  feeCurrency: Address
) {
  const l2 = await isCel2(ctx);

  const address = await getContractAddress(
    ctx,
    l2 ? "FeeCurrencyDirectory" : "FeeCurrencyWhitelist"
  );

  const contract = new Contract(
    l2 ? feeCurrencyDirectoryABI : feeCurrencyWhitelistABI,
    address
  );
  contract.link(ctx);

  const whitelist = await contract.methods[
    l2 ? "getCurrencies" : "getWhitelist"
  ]().call();

  return whitelist
    .map((x) => x.toLowerCase())
    .includes(feeCurrency.toLowerCase());
}
