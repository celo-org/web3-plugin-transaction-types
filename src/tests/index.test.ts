import { describe, beforeAll, expect, test, beforeEach } from "bun:test";
import Web3 from "web3";

import { CeloTransactionTypesPlugin } from "../index";
import { CeloChains } from "../utils";

describe("CeloTransactionTypesPlugin", () => {
  let web3: Web3;

  beforeAll(() => {
    web3 = new Web3();
    web3.registerPlugin(new CeloTransactionTypesPlugin());
    web3.celo.link(web3);
  });
  beforeEach(() => {
    web3.setProvider(CeloChains.alfajores.rpcUrl);
  });

  test("should be registered under .celo namespace", () => {
    expect(web3.celo).toMatchSnapshot();
  });

  describe("isValidFeeCurrency()", () => {
    test("l1", async () => {
      expect(web3.celo.isValidFeeCurrency("0x123")).resolves.toBe(false);
      expect(
        web3.celo.isValidFeeCurrency(
          "0x874069Fa1Eb16D44d622F2e0Ca25eeA172369bC1"
        )
      ).resolves.toBe(true);
    });
    test("l2", async () => {
      web3.setProvider(CeloChains.dango.rpcUrl);
      expect(web3.celo.isValidFeeCurrency("0x123")).resolves.toBe(false);
      expect(
        web3.celo.isValidFeeCurrency(
          "0x4822e58de6f5e485eF90df51C41CE01721331dC0"
        )
      ).resolves.toBe(true);
    });
  });
});
