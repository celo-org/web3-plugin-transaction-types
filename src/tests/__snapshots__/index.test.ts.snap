// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`CeloTransactionTypesPlugin should be registered under .celo namespace 1`] = `
CeloTransactionTypesPlugin {
  "_accountProvider": {
    "create": [Function],
    "decrypt": [Function],
    "encrypt": [Function],
    "hashMessage": [Function],
    "parseAndValidatePrivateKey": [Function],
    "privateKeyToAccount": [Function],
    "privateKeyToAddress": [Function],
    "privateKeyToPublicKey": [Function],
    "recover": [Function],
    "recoverTransaction": [Function],
    "sign": [Function],
    "signTransaction": [Function],
    "wallet": [],
  },
  "_emitter": EventEmitter {
    "_events": {
      "CONFIG_CHANGE": [
        EE {
          "context": [Circular],
          "fn": [Function],
          "once": false,
        },
        EE {
          "context": [Circular],
          "fn": [Function],
          "once": false,
        },
        EE {
          "context": [Circular],
          "fn": [Function],
          "once": false,
        },
        EE {
          "context": [Circular],
          "fn": [Function],
          "once": false,
        },
        EE {
          "context": [Circular],
          "fn": [Function],
          "once": false,
        },
        EE {
          "context": [Circular],
          "fn": [Function],
          "once": false,
        },
        EE {
          "context": [Circular],
          "fn": [Function],
          "once": false,
        },
        EE {
          "context": [Circular],
          "fn": [Function],
          "once": false,
        },
      ],
    },
    "_eventsCount": 1,
    "maxListeners": 9007199254740991,
  },
  "_requestManager": Web3RequestManager {
    "_emitter": EventEmitter {
      "_events": {
        "BEFORE_PROVIDER_CHANGE": EE {
          "context": [Circular],
          "fn": [Function],
          "once": false,
        },
        "PROVIDER_CHANGED": EE {
          "context": [Circular],
          "fn": [Function],
          "once": false,
        },
      },
      "_eventsCount": 2,
      "maxListeners": 9007199254740991,
    },
    "_provider": HttpProvider {
      "clientUrl": "https://alfajores-forno.celo-testnet.org",
      "httpProviderOptions": undefined,
    },
    "useRpcCallSpecification": undefined,
  },
  "_subscriptionManager": Web3SubscriptionManager {
    "_subscriptions": Map {},
    "registeredSubscriptions": {
      "logs": [class LogsSubscription],
      "newBlockHeaders": [class NewHeadsSubscription],
      "newHeads": [class NewHeadsSubscription],
      "newPendingTransactions": [class NewPendingTransactionsSubscription],
      "pendingTransactions": [class NewPendingTransactionsSubscription],
      "syncing": [class SyncingSubscription],
    },
    "requestManager": Web3RequestManager {
      "_emitter": EventEmitter {
        "_events": {
          "BEFORE_PROVIDER_CHANGE": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
          "PROVIDER_CHANGED": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
        },
        "_eventsCount": 2,
        "maxListeners": 9007199254740991,
      },
      "_provider": HttpProvider {
        "clientUrl": "https://alfajores-forno.celo-testnet.org",
        "httpProviderOptions": undefined,
      },
      "useRpcCallSpecification": undefined,
    },
    "tolerateUnlinkedSubscription": false,
  },
  "_wallet": [],
  "config": {
    "blockHeaderTimeout": 10,
    "contractDataInputFill": "data",
    "defaultAccount": undefined,
    "defaultBlock": "latest",
    "defaultChain": "mainnet",
    "defaultCommon": undefined,
    "defaultHardfork": "london",
    "defaultMaxPriorityFeePerGas": "0x9502f900",
    "defaultNetworkId": undefined,
    "defaultReturnFormat": {
      "bytes": "BYTES_HEX",
      "number": "NUMBER_BIGINT",
    },
    "defaultTransactionType": "0x2",
    "enableExperimentalFeatures": {
      "useRpcCallSpecification": false,
      "useSubscriptionWhenCheckingBlockTimeout": false,
    },
    "handleRevert": false,
    "maxListenersWarningThreshold": 100,
    "transactionBlockTimeout": 50,
    "transactionBuilder": undefined,
    "transactionConfirmationBlocks": 24,
    "transactionConfirmationPollingInterval": undefined,
    "transactionPollingInterval": 1000,
    "transactionPollingTimeout": 750000,
    "transactionReceiptPollingInterval": undefined,
    "transactionSendTimeout": 750000,
    "transactionTypeParser": undefined,
  },
  "pluginNamespace": "celo",
  "providers": {
    "HttpProvider": [class HttpProvider],
    "WebsocketProvider": [class WebSocketProvider],
  },
}
`;
